astropy/io/fits/header.py

```diff
@@ -2205,11 +2205,7 @@ def __eq__(self, other):
             else:
                 return False
 
-        for a, b in itertools.zip_longest(self, other):
-            if a != b:
-                return False
-        else:
-            return True
+        return all(not a != b for a, b in itertools.zip_longest(self, other))
```

Comment:
At first glance using `all` and "list comprehension" is more compact and I like it because of that. However, the original code is more efficient since it can exit the look even after the first element is tested. The list comprehension needs to loop over the entire thing. I prefer more efficient code.

Reply:
Actually python internally optimizes and `all` will bail when it meets a counterexample. So the comprehension is both cleaner and faster than the for loop.

Reply:
`any` works similarly. It will exit when it mess the first positive example.

Reply:
For reference, this is how python does the optimization: https://en.wikipedia.org/wiki/Short-circuit_evaluation

Reply:
```suggestion
        return all(a == b for a, b in itertools.zip_longest(self, other))
```
And it is worth pointing out that there is no [list comprehension](https://docs.python.org/3/glossary.html#term-list-comprehension) here, instead there is a [generator expression](https://docs.python.org/3/glossary.html#term-generator-expression).